core/modules/shrscan/Director.h

```diff
@@ -0,0 +1,315 @@
+//!
+//! @return false: Initialization failed (e.g. not all chunks were found).
+//! @return true:  Initialization succeeded.
+//-----------------------------------------------------------------------------
+
+virtual bool  init(const std::string dbPath, TableInfo const& baseTables,
+                   std::shared_ptr<std::vector<int>> const& chunkList);
+
+//-----------------------------------------------------------------------------
+//! @brief Resume a possibly suspended shared scan query.
+//!
+//! When a query is resumed it becomes eligible to run when sufficient
+//! resources become available.
+//!
+//! @param  query - Pointer to a Query object that describes the query.
+//!
+//! @return false: The query was not found.
+//! @return true:  The query was resumed.
+//-----------------------------------------------------------------------------
+
+virtual bool  resumeQuery(Query const* query) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Resume all suspended shared scan queries.
+//!
+//! This method effectively calls resumeQuery() on each suspended query.
+//-----------------------------------------------------------------------------
+
+virtual void  resumeAll() = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Set the maximum amount of memory that may be locked.
+//!
+//! The default maximum amount of memory that may be locked is equal to largest
+//! amount of memory needed to lock a set of table chunks passed to init().
+//!
+//! @param  bytes  - The maximum number of bytes that may be locked.
+//-----------------------------------------------------------------------------
+
+virtual void  setMaxMemory(unsigned long long bytes) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Set the maximum number of queries that be run in parallel.
+//!
+//! The default number of parallel queries is 7.
+//!
+//! @param  maxNum - The maximum number of queries to run in parallel.
+//-----------------------------------------------------------------------------
+
+virtual void  SetMaxQueries(int qnum) = 0;
+
+//-----------------------------------------------------------------------------
+//! @brief Suspend a shared scan query.
+//!
+//! When a running query is suspended, it is placed in the eligible to run
+//! queue in FIFO order. If the query is executing its runQuery() method, the
+//! suspension occurs upon return.
+//!
+//! @param  query   - Pointer to the Query object that describes the query.
+//! @param  unlock  - When true, any additional tables required by the query
+//!                   that were locked are unlocked. Otherwise, the tables
+//!                   remain locked which may prevent new queries to run.
+//!
+//! @return false: The query was not found.
+//! @return true:  The query was suspended or scheduled to be suspended.
```

Comment:
I don't like throwing exceptions for things that are not really 
exceptions. In this case, the query could have been cancelled (either by 
itself or someone else) or finished. So, not finding a query is not an 
exception. Returning a true/false value allows the caller to do something 
reasonable in an easy way. I wish there was a standard way of warning of 
not using a returned value which would be more effective here.

Andy

On Wed, 18 Nov 2015, Andy wrote:

>> +virtual void  SetMaxQueries(int qnum) = 0;
>> +
>> +//-----------------------------------------------------------------------------
>> +//! @brief Suspend a shared scan query.
>> +//!
>> +//! When a running query is suspended, it is placed in the eligible to run
>> +//! queue in FIFO order. If the query is executing its runQuery() method, the
>> +//! suspension occurs upon return.
>> +//!
>> +//! @param  query   - Pointer to the Query object that describes the query.
>> +//! @param  unlock  - When true, any additional tables required by the query
>> +//!                   that were locked are unlocked. Otherwise, the tables
>> +//!                   remain locked which may prevent new queries to run.
>> +//!
>> +//! @return false: The query was not found.
>> +//! @return true:  The query was suspended or scheduled to be suspended.
>
> Should probably raise an exception if query is not found.
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/lsst/qserv/pull/156/files#r45247711